proposals/left-right-join-in-query-expressions.md

```diff
@@ -0,0 +1,235 @@
+
+## Detailed design
+[design]: #detailed-design
+
+Proposed grammar change ([§11.7.1](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1117-query-expressions)):
+
+### Grammar
+
+```diff
+join_clause
+-    : 'join' type? identifier 'in' expression 'on' expression
++    : ('left' | 'right')? 'join' type? identifier 'in' expression 'on' expression
+      'equals' expression
+    ;
+```
+
+As per the grammar proposal, the syntax for the proposed `left join` and `right join` expression clauses would be identical to the existing `join` clause. For example:
+
+```c#
+from student in students
+left join department in departments on student.DepartmentID equals department.ID
+select new { student.Name, department?.Name }
+```
+
+### Left join translation
+
+A `left join` clause immediately followed by a `select` clause
+
+```csharp
+from «x1» in «e1»  
+left join «x2» in «e2» on «k1» equals «k2»  
+select «v»
+```
+
+is translated into
+
+```csharp
+( «e1» ) . LeftJoin( «e2» , «x1» => «k1» , «x2» => «k2» , ( «x1» , «x2» ) => «v» )
+```
+
+> *Example*: The example
+>
+> ```csharp
+> from c in customersh
+> left join o in orders on c.CustomerID equals o.CustomerID
+> select new { c.Name, o?.OrderDate, o?.Total }
+> ```
+>
+> is translated into
+>
+> ```csharp
+> (customers).LeftJoin(
+>    orders,
+>    c => c.CustomerID, o => o.CustomerID,
+>    (c, o) => new { c.Name, o?.OrderDate, o?.Total })
+> ```
+>
+> *end example*
+
+A `left join` clause followed by a query body clause:
+
+```csharp
+from «x1» in «e1»  
+left join «x2» in «e2» on «k1» equals «k2»  
+...
+```
+
+is translated into
+
+```csharp
+from * in ( «e1» ) . LeftJoin(  
+«e2» , «x1» => «k1» , «x2» => «k2» ,
+( «x1» , «x2» ) => new { «x1» , «x2» })  
+...
+```
+
+### Right join translation
+
+A `right join` clause immediately followed by a `select` clause
+
+```csharp
+from «x1» in «e1»  
+right join «x2» in «e2» on «k1» equals «k2»  
+select «v»
+```
+
+is translated into
+
+```csharp
+( «e1» ) . RightJoin( «e2» , «x1» => «k1» , «x2» => «k2» , ( «x1» , «x2» ) => «v» )
+```
+
+> *Example*: The example
+>
+> ```csharp
+> from o in orders
+> right join c in customersh on o.CustomerID equals c.CustomerID
+> select new { o?.OrderDate, o?.Total, c.Name }
+> ```
+>
+> is translated into
+>
+> ```csharp
+> (orders).RightJoin(
+>    customers,
+>    o => o.CustomerID, c => c.CustomerID,
+>    (c, o) => new { o?.OrderDate, o?.Total, c.Name })
+> ```
+>
+> *end example*
+
+A `right join` clause followed by a query body clause:
+
+```csharp
+from «x1» in «e1»  
+right join «x2» in «e2» on «k1» equals «k2»  
+...
+```
+
+is translated into
+
+```csharp
+from * in ( «e1» ) . RightJoin(  
+«e2» , «x1» => «k1» , «x2» => «k2» ,
+( «x1» , «x2» ) => new { «x1» , «x2» })  
+...
+```
```

Comment:
It might be simpler to instead just add a section to the spec that says that for `join clauses` that the existing translation stays the same *except* that if the `join_clause` has a `left or right` keyword, that we call translate into LeftJoin or RightJoin respectively.  That would *greatly* minimize the size of the spec change being proposed.